format/Flight.proto

```diff
@@ -525,3 +525,108 @@ message FlightData {
 message PutResult {
   bytes app_metadata = 1;
 }
+
+/*
+ * EXPERIMENTAL: Union of possible value types for a Session Option to be set to.
+ */
+message SessionOptionValue {
+  message StringListValue {
+    repeated string values = 1;
+  }
+
+  oneof option_value {
+    string string_value = 1;
+    bool bool_value = 2;
+    sfixed32 int32_value = 3;
+    sfixed64 int64_value = 4;
+    float float_value = 5;
+    double double_value = 6;
+    StringListValue string_list_value = 7;
+  }
+}
+
+/*
+ * EXPERIMENTAL: A request to set session options for an existing or new (implicit)
+ * server session.
+ *
+ * Sessions are persisted and referenced via RFC 6265 HTTP cookies, canonically
+ * 'arrow_flight_session_id', although implementations may freely choose their own name.
+ */
+message SetSessionOptionsRequest {
+  map<string, SessionOptionValue> session_options = 1;
+}
+
+/*
+ * EXPERIMENTAL: The results (individually) of setting a set of session options.
+ */
+message SetSessionOptionsResult {
+  enum Status {
+    // The status of setting the option is unknown. Servers should avoid using
+    // this value (send a NOT_FOUND error if the requested query is
+    // not known). Clients can retry the request.
+    SET_SESSION_OPTION_RESULT_UNSPECIFIED = 0;
+    // The session option setting completed successfully.
+    SET_SESSION_OPTION_RESULT_OK = 1;
+    // The given session option name was an alias for another option name.
+    SET_SESSION_OPTION_RESULT_OK_MAPPED = 2;
+    // The given session option name is invalid.
+    SET_SESSION_OPTION_RESULT_INVALID_NAME = 3;
+    // The session option value is invalid.
+    SET_SESSION_OPTION_RESULT_INVALID_VALUE = 4;
+    // The session option cannot be set.
+    SET_SESSION_OPTION_RESULT_ERROR = 5;
+  }
+
+  message Result {
+    Status status = 1;
+  }
+  
+  map<string, Result> results = 1;
```

Comment:
Isn't this a bit overkill? Do we envision clients doing different things based on per-key results?

Reply:
Certainly.  For example to return that an option value cannot be set on the server, we either expose this or construct a structured error message to capture which option names had which success or error.  This way, drivers or generic clients could easily (and programmatically, without parsing human-readable error/warning text) establish a fallback behaviour given unavailability of specific keys or values.  In a more simple case, some specific name or value failures may be fatal while others may be attempts to set non-critical "preferences" e.g. more efficient result packing or the like.

Reply:
Would it be reasonable for the "default" to be success, and just *only* report failures/errors, and the client can assume any option they set which is not reported in the response was successful?

Reply:
> Would it be reasonable for the "default" to be success, and just _only_ report failures/errors, and the client can assume any option they set which is not reported in the response was successful?

Unfortunately as @lidavidm discussed elsewhere the default/0 enum value is used by Proto3 (in some languages—I believe C++ and not Java?) as a fallback for unrecognized packed values, so we cannot use it as a meaningful value.  It would certainly be nice to check the enum is <= OK_MAPPED or whatnot, rather than a double bounds check, but we can't...

Reply:
> Would it be reasonable for the "default" to be success, and just _only_ report failures/errors, and the client can assume any option they set which is not reported in the response was successful?

I would presumably have to eliminate the OK_MAPPED status, but yes this would keep the parsing consistent with the request structure AND eliminate a lot of work for the client by reducing the MVP to checking the length of the response map.

Reply:
> Should we reserve the OK statuses in case for some hitherto-unknown reason we decide to use them eventually (so that they're consistent with the enum ordering used elsewhere)?

This seems overkill.

Reply:
> > Would it be reasonable for the "default" to be success, and just _only_ report failures/errors, and the client can assume any option they set which is not reported in the response was successful?
> 
> I would presumably have to eliminate the OK_MAPPED status, but yes this would keep the parsing consistent with the request structure AND eliminate a lot of work for the client by reducing the MVP to checking the length of the response map.

@pitrou does this option satisfy your concerns here, as it seems to me to?

Reply:
> @pitrou does this option satisfy your concerns here, as it seems to me to?

Yes, it would!

Reply:
Done in Proto & C++, WIP in Java.  Thanks @zeroshade @pitrou